cli/src/pcluster/models/cluster.py

```diff
@@ -345,63 +356,25 @@ def _validate_and_parse_config(self, suppress_validators, validation_failure_lev
 
         return config
 
-    def _setup_cluster_bucket(self) -> ClusterBucket:
-        """
-        Create pcluster bucket, if needed.
-
-        If no bucket specified, create a bucket associated to the given stack.
-        Created bucket needs to be removed on cluster deletion.
-        """
-        if self.config.cluster_s3_bucket:
-            # Use user-provided bucket
-            # Do not remove this bucket on deletion, but cleanup artifact directory
-            name = self.config.cluster_s3_bucket
-            remove_on_deletion = False
-            try:
-                check_s3_bucket_exists(name)
-            except Exception as e:
-                LOGGER.error("Unable to access config-specified S3 bucket %s: %s", name, e)
-                raise e
-        else:
-            # Create 1 bucket per cluster named "parallelcluster-{random_string}" if bucket is not provided
-            # This bucket needs to be removed on cluster deletion
-            name = generate_random_name_with_prefix("parallelcluster")
-            # self.cluster_s3_bucket = name
-            remove_on_deletion = True
-            LOGGER.debug("Creating S3 bucket for cluster resources, named %s", name)
-            try:
-                create_s3_bucket(name)
-            except Exception as e:
-                LOGGER.error("Unable to create S3 bucket %s.", name)
-                raise e
-
-        # Use "{stack_name}-{random_string}" as directory in bucket
-        artifact_directory = generate_random_name_with_prefix(self.stack_name)
-        return ClusterBucket(name, artifact_directory, remove_on_deletion)
-
     def _upload_config(self):
         """Upload source config and save config version."""
         if not self.bucket:
             ClusterActionError("S3 bucket must be created before uploading artifacts.")
 
         try:
-            # Upload original config
-            if self.config.source_config:
-                AWSApi.instance().s3.put_object(
-                    bucket_name=self.bucket.name,
-                    body=yaml.dump(self.config.source_config),
-                    key=self._get_original_config_key(),
-                )
             # Upload config with default values and sections
             if self.config:
-                result = AWSApi.instance().s3.put_object(
-                    bucket_name=self.bucket.name,
-                    body=yaml.dump(ClusterSchema().dump(self.config)),
-                    key=self._get_config_key(),
+                result = self.bucket.upload_config(
+                    config=ClusterSchema().dump(self.config), config_name=self.bucket.config_name
                 )
+
                 # config version will be stored in DB by the cookbook at the first update
                 self.config.config_version = result.get("VersionId")
 
+            # Upload original config
+            if self.config.source_config:
```

Comment:
yes I'd avoid all the extra conditional checks that are not really needed so that we simplify the code. Otherwise we end up having this if conditionals everywhere in the code. It is good to use a defensive approach and check preconditions on methods that represent a library interface or functions that are used in modules we have poor control over but in the other cases we can avoid it. In this case if source_config is None it is more of a bug I'd say.